src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenRefReadonlyParameterTests.cs

```diff
@@ -0,0 +1,595 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics;
+
+// PROTOTYPE: Move to RefReadonlyParameterTests.cs.
+public partial class RefReadonlyParameterTests : CSharpTestBase
+{
+    private const string RequiresLocationAttributeName = "RequiresLocationAttribute";
+    private const string RequiresLocationAttributeNamespace = "System.Runtime.CompilerServices";
+    private const string RequiresLocationAttributeQualifiedName = $"{RequiresLocationAttributeNamespace}.{RequiresLocationAttributeName}";
+
+    private static void VerifyRequiresLocationAttributeSynthesized(ModuleSymbol module)
+    {
+        var attributeType = module.GlobalNamespace.GetMember<NamedTypeSymbol>(RequiresLocationAttributeQualifiedName);
+        if (module is SourceModuleSymbol)
+        {
+            Assert.Null(attributeType);
+        }
+        else
+        {
+            Assert.NotNull(attributeType);
+        }
+    }
+
+    private static void VerifyRefReadonlyParameter(ParameterSymbol parameter,
+        bool refKind = true,
+        bool metadataIn = true,
+        bool attributes = true,
+        bool modreq = false,
+        bool noUseSiteErrors = true)
+    {
+        if (refKind)
+        {
+            Assert.Equal(RefKind.RefReadOnlyParameter, parameter.RefKind);
+        }
+
+        if (metadataIn)
+        {
+            Assert.True(parameter.IsMetadataIn);
+        }
+
+        if (attributes)
+        {
+            if (parameter.ContainingModule is SourceModuleSymbol)
+            {
+                Assert.Empty(parameter.GetAttributes());
+            }
+            else
+            {
+                var attribute = Assert.Single(parameter.GetAttributes());
+                Assert.Equal("System.Runtime.CompilerServices.RequiresLocationAttribute", attribute.AttributeClass.ToTestDisplayString());
+                Assert.Empty(attribute.ConstructorArguments);
+                Assert.Empty(attribute.NamedArguments);
+            }
+        }
+
+        if (modreq)
+        {
+            var mod = Assert.Single(parameter.RefCustomModifiers);
+            Assert.Equal("System.Runtime.InteropServices.InAttribute", mod.Modifier.ToTestDisplayString());
+        }
+        else
+        {
+            Assert.Empty(parameter.RefCustomModifiers);
+        }
+
+        var method = (MethodSymbol)parameter.ContainingSymbol;
+
+        if (noUseSiteErrors)
+        {
+            Assert.False(method.HasUnsupportedMetadata);
+            Assert.False(method.HasUseSiteError);
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

It feels like we are missing the ```else``` part and consumers have to check presence of errors themselves. Also, ``` noUseSiteErrors: false``` is a double negative. Consider reverting the meaning of the parameter and renaming.